dynawo/sources/Models/CPP/Controls/Voltage/ModelVoltageMeasurementsUtilities/DYNModelVoltageMeasurementsUtilities.cpp

```diff
@@ -0,0 +1,343 @@
+ *
+ * @return A pointer to a new instance of ModelVoltageMeasurementsUtilitiesFactory
+ */
+extern "C" DYN::SubModelFactory* getFactory() {
+  return (new DYN::ModelVoltageMeasurementsUtilitiesFactory());
+}
+
+/**
+ * @brief ModelVoltageMeasurementsUtilitiesFactory destroy method
+ */
+extern "C" void deleteFactory(DYN::SubModelFactory* factory) {
+  delete factory;
+}
+
+/**
+ * @brief ModelVoltageMeasurementsUtilities getter
+ *
+ * @return A pointer to a new instance of ModelVoltageMeasurementsUtilities
+ */
+extern "C" DYN::SubModel* DYN::ModelVoltageMeasurementsUtilitiesFactory::create() const {
+  DYN::SubModel* model(new DYN::ModelVoltageMeasurementsUtilities());
+  return model;
+}
+
+/**
+ * @brief ModelVoltageMeasurementsUtilitiesFactory destroy method
+ */
+extern "C" void DYN::ModelVoltageMeasurementsUtilitiesFactory::destroy(DYN::SubModel* model) const {
+  delete model;
+}
+
+namespace DYN {
+
+/**
+ * @brief ModelVoltageMeasurementsUtilities model default constructor
+ *
+ */
+ModelVoltageMeasurementsUtilities::ModelVoltageMeasurementsUtilities() :
+ModelCPP("voltageMeasurementsUtilities"),
+nbConnectedInputs_(0) {
+}
+
+/**
+ * @brief ModelVoltageMeasurementsUtilities model initialization
+ *
+ */
+void
+ModelVoltageMeasurementsUtilities::init(const double /*t0*/) {
+  // Nothing to initalize here.
+}
+
+void
+ModelVoltageMeasurementsUtilities::initializeFromData(const boost::shared_ptr<DataInterface>& /*data*/) {
+  // not needed
+}
+
+void
+ModelVoltageMeasurementsUtilities::getSize() {
+  sizeF_ = 0;  // No equartions
+  sizeY_ = nbConnectedInputs_;  // All voltage inputs and their boolean activity values
+  sizeZ_ = nbConnectedInputs_;
+  sizeG_ = 0;
+  sizeMode_ = 0;
+
+  calculatedVars_.assign(nbCalculatedVars_, 0);
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalF(double /*t*/, propertyF_t /*type*/) {
+  // No evalF function needed
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalG(const double /*t*/) {
+  // No root function for this model
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalJt(const double /*t*/, const double /*cj*/, SparseMatrix& /*jt*/, const int /*rowOffset*/) {
+  // No evalJt function needed
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalJtPrim(const double /*t*/, const double /*cj*/, SparseMatrix& /*jt*/, const int /*rowOffset*/) {
+  // No evalJtPrim function needed
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalZ(const double /*t*/) {
+  // No evalZ function needed
+}
+
+void
+ModelVoltageMeasurementsUtilities::collectSilentZ(BitMask* silentZTable) {
+  for (unsigned int s = 0; s < sizeZ_; ++s) {
+    silentZTable[s].setFlags(NotUsedInContinuousEquations);
+  }
+}
+
+modeChangeType_t
+ModelVoltageMeasurementsUtilities::evalMode(const double /*t*/) {
+  return NO_MODE;
+}
+
+void
+ModelVoltageMeasurementsUtilities::evalJCalculatedVarI(unsigned /*iCalculatedVar*/, vector<double>& /*res*/) const {
+  // output depends only on discrete variables
```

Comment:
right! we can go with a step by step implementation. The writing of the jacobian should be quite easy. The point to take also into account is that, in case there is an abrupt change of voltage of one of the connected components, a mode should be raised to notify a new shape for the jacobian (e.g. 1 appearing somewhere else on a column/row that what before). If a component is disconnected hopefully, for a first implementation, a mode is raised somewhere else and the jacobian is recalculated

Reply:
There are a couple of options. The min / max being singular, it is worth looking at various options: here are a few I can think of: 
1. Run through the list of inputs to find the first one which reaches the min / max. Set this one entry to 1 and the others to 0. (current implementation, not yet pushed) 
2. Run through the list of inputs to find all the indices reaching the min / max. Set the associated entries to 1/number of non zero and the other ones to 0. 
3. Approximate the min / max with lp norms (say p =10 for the max and -10 for the min?). The advantage here is that it allows to compute fairly accurate approximations yet we do not need to evaluate the mode changes. 

Your choice is mine :) 
@rosiereflo @marcochiaramello

Reply:
We went with a discrete model for computing everything, this should no longer be an issue.